packages/ERTP/src/eventualWithdraw.js

```diff
@@ -0,0 +1,40 @@
+// @ts-check
+import { E } from '@agoric/eventual-send';
+import { makePromiseKit } from '@agoric/promise-kit';
+
+import './types';
+
+/** @type {EventualWithdraw} */
+export const eventualWithdraw = (
+  purse,
+  amountMath,
+  searchAmount,
+  iterations = 1,
+) => {
+  const notifier = E(purse).getCurrentAmountNotifier();
+  const paymentPromiseKit = makePromiseKit();
+
+  const tryToFind = async lastUpdateCount => {
+    if (iterations <= 0) {
+      paymentPromiseKit.reject(Error('searchAmount could not be withdrawn'));
+      return;
+    }
+    const { value: currentAmount, updateCount } = await E(
+      notifier,
+    ).getUpdateSince(lastUpdateCount);
+    const foundAmount = amountMath.find(currentAmount, searchAmount);
+
+    if (amountMath.isEmpty(foundAmount)) {
+      // searchAmount was not found, try again
+      iterations -= 1;
+      tryToFind(updateCount);
+      return;
+    }
+
+    // searchAmount was found! Withdraw a payment
+    paymentPromiseKit.resolve(E(purse).withdraw(foundAmount));
```

Comment:
Because of the unavoidable race condition, the `withdraw` may fail even though the iterations are not exhausted. Since the loop is eventual anyway, if the `withdraw` fails I think the loop should continue. (Assuming the `iterations` thing is a good idea, which I don't yet have an opinion about.)

Reply:
See https://github.com/Agoric/agoric-sdk/pull/2155#pullrequestreview-560256142

In the set case, the `foundAmount` obtained at line 25 may be larger than the assets remaining in the purse at line 35, causing the withdraw to fail. However, at that moment the purse may actually have enough assets to satisfy the pattern, so it is weird that those "enough" assets were not withdrawn.

Reply:
Would a set intersection operation (which could then be checked against isEmpty) be useful?

Reply:
How so?